api/v1alpha1/runtime.proto

```diff
@@ -38,16 +38,34 @@ service ControllerAdapter {
 	rpc RemoveFinalizer(RuntimeRemoveFinalizerRequest) returns (RuntimeRemoveFinalizerResponse);
 }
 
-// RegisterController RPC
+enum ControllerInputKind {
+    WEAK = 0;
+    STRONG = 1;
+}
 
-message RegisterControllerRequest {
-    string controller_name = 1;
-    ManagedResources managed_resources = 2;
+message ControllerInput {
+    ControllerInputKind kind = 1;
+    string namespace = 2;
+    string type = 3;
+    optional string id = 4;
+}
+
+enum ControllerOutputKind {
+    EXCLUSIVE = 0;
+    SHARED = 1;
 }
 
-message ManagedResources {
-    string namespace = 1;
+message ControllerOutput {
     string type = 2;
+    ControllerOutputKind kind = 3;
+}
+
+// RegisterController RPC
+
+message RegisterControllerRequest {
+    string controller_name = 1;
+    repeated ControllerInput initial_inputs = 3;
```

Comment:
My opinion, so take with a grain of salt: I think this can be simply `inputs`. To me it is clear that I can register with inputs and then update them after the fact. Calling them the same thing (just `inputs`) in each RPC solidifies that concept for me a bit more.

Reply:
yes, I was torn on this one, I agree `Inputs/Outputs` sounds easier